### DISCUSSION TITLE
Bluetooth boombox

### DISCUSSION THREAD

**Discussion by Alfred586** [pschatzmann/ESP32-A2DP#122] [Show and tell]
The power of two cores!
Thanks for your support @pschatzmann 
Sink mode, avrcp commands, metadata and fast fourier transform
![20211113_183649](https://user-images.githubusercontent.com/88184503/141663215-f1f2fdfd-3d55-4497-aec0-e1d65a0b33f4.jpg)

**Comment by tankhunter8192** [comment]
nice build, what kind of i2s module did you use for that?

**Reply by Alfred586** [reply]
Thanks.
It is a custom board made by myself.
Contains a PCM5102 DAC, Esp32, and a raspberry pi display.

**Comment by RWBRWB** [comment]
What code did you use for the FFT? does that work on Web Audio Streams also?

**Reply by Alfred586** [reply]
Hello.
arduinoFFT library, and yes it works whith any class of signal you put in
You only have to fill the fft Real array with your music data.

For example:

int m = 0;
void read_data_stream(const uint8_t *data_stream, uint32_t len)  // Function to acquire the music data
{
  int16_t *values = (int16_t*) data_stream;
  int m = 0;
    for (int j=0; j<len/2; j+=2){
      mBuffer[m] = ((values[j] + values[j+1])/2);  // Mix the two channels 
     m += 1;
    }  
}

for (int i = 0; i < SAMPLES; i++) {
    newTime = micros();
    vReal[i] = map(mBuffer[i], -32767, 32767, 0, 4096); // Fill the real data buffer
    vImag[i] = 0;

   while ((micros() - newTime) < sampling_period_us) {
      // do nothing to wait
      yield();
    }
  }

Cheers.

**Comment by pschatzmann** [comment]
I would also suggest that you have a look at the FFT functionality provided by the [Arduino Audio Tools](https://github.com/pschatzmann/arduino-audio-tools). There this is just one of the many Audio Sinks that you can write or copy the audio data to...
See also https://github.com/pschatzmann/arduino-audio-tools/issues/240

**Comment by msaraswat15031997** [comment]
Wow, this is all that I need. I am working on a similar project.  I need to do all that you've done and I'm a beginner . Can I please see understand and learn from your code. Much appreciated   thanks 🙏
